File: client/src/net.ts
/**
 * WebSocket client with auto-reconnect. Encrypts every outbound update and
 * decrypts inbound peer blobs. The socket carries only opaque ciphertext plus
 * the routing roomId; the AES key stays in this process.
 */
import { decryptJson, encryptJson, type RoomKeys } from "./crypto.js";
import type { ServerMessage } from "../../shared/messages.js";
import type { PeerUpdate } from "./types.js";

export interface NetHandlers {
  onPeer: (id: string, update: PeerUpdate) => void;
  onLeft: (id: string) => void;
  onRequest: () => void; // a peer joined; re-broadcast our latest state
  onStatus: (connected: boolean) => void;
  onPresence: (n: number) => void; // room occupancy (count only, no identity)
  onFatal: (reason: string) => void;
}

export class NetClient {
  private ws: WebSocket | null = null;
  private backoff = 500;
  private closed = false;
  private lastResync = 0;
  private lastSent: string | null = null; // for latest-state re-broadcast on request

  constructor(private readonly keys: RoomKeys, private readonly h: NetHandlers) {}

  connect(): void {
    // Always start from a clean slate. A previous socket may be dead-but-OPEN
    // (mobile standby zombie) or wedged in CONNECTING (network handover on wake);
    // tearing it down first means a reconnect attempt never stacks a second socket
    // on top of a live one, nor gets blocked waiting on a hung one.
    this.teardown();
    const proto = location.protocol === "https:" ? "wss" : "ws";
    const ws = new WebSocket(`${proto}://${location.host}/ws`);
    this.ws = ws;

    ws.onopen = () => {
      this.backoff = 500;
      ws.send(JSON.stringify({ t: "join", roomId: this.keys.roomId }));
      this.h.onStatus(true);
      if (this.lastSent) ws.send(this.lastSent); // resume visibility after reconnect
    };

    ws.onmessage = async (ev) => {
      let msg: ServerMessage;
      try {
        msg = JSON.parse(ev.data as string);
      } catch {
        return;
      }
      if (msg.t === "peer") {
        const update = await decryptJson<PeerUpdate>(this.keys.key, msg.data);
        if (update) this.h.onPeer(msg.id, update);
      } else if (msg.t === "left") {
        this.h.onLeft(msg.id);
      } else if (msg.t === "request") {
        this.h.onRequest();
      } else if (msg.t === "presence") {
        this.h.onPresence(msg.n);
      } else if (msg.t === "error") {
        this.closed = true;
        this.h.onFatal(msg.reason);
      }
    };

    ws.onclose = () => {
      this.h.onStatus(false);
      if (!this.closed) this.scheduleReconnect();
    };
    ws.onerror = () => ws.close();
  }

  /** Detach and close the current socket without triggering its handlers. */
  private teardown(): void {
    const ws = this.ws;
    if (!ws) return;
    ws.onopen = null;
    ws.onmessage = null;
    ws.onerror = null;
    ws.onclose = null;
    try {
      ws.close();
    } catch {
      /* ignore */
    }
    this.ws = null;
  }

  private scheduleReconnect(): void {
    const delay = Math.min(this.backoff, 15_000);
    this.backoff = Math.min(this.backoff * 2, 15_000);
    setTimeout(() => {
      if (!this.closed) this.connect();
    }, delay);
  }

  /**
   * Force a fresh connection. Call this when the page returns to the foreground:
   * standby/tab-switch often leaves a "zombie" socket that still reports OPEN but
   * carries nothing, so we can't rely on onclose. Reconnecting makes the server
   * re-request peer broadcasts (so we see them) and re-sends our last position
   * (so they see us).
   */
  resync(): void {
    if (this.closed) return;
    const now = Date.now();
    if (now - this.lastResync < 2000) return; // debounce rapid focus/blur
    this.lastResync = now;
    this.backoff = 500;
    this.connect(); // connect() tears down any existing socket first
  }

  async broadcast(update: PeerUpdate): Promise<void> {
    const data = await encryptJson(this.keys.key, update);
    const frame = JSON.stringify({ t: "relay", data });
    // lastSent is replayed on every reconnect (see onopen). Keep it only while we are
    // actively sharing a position; a "stop" must clear it, or a later reconnect would
    // resurrect our old position and make us reappear to peers after we stopped.
    if (update.k === "loc") this.lastSent = frame;
    else if (update.k === "stop") this.lastSent = null;
    if (this.ws && this.ws.readyState === WebSocket.OPEN) this.ws.send(frame);
  }

  close(): void {
    this.closed = true;
    this.ws?.close();
  }
}

File: mobile/lib/core/net.dart
/// WebSocket client with auto-reconnect. Port of `client/src/net.ts`.
///
/// Encrypts every outbound update and decrypts inbound peer blobs. The socket
/// carries only opaque ciphertext plus the routing roomId; the AES key stays in
/// this process. Timings are fixed by `shared/PROTOCOL.md` — changing one here
/// alone makes this client and the browser time each other out inconsistently.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'crypto.dart';
import 'protocol.dart';
import 'types.dart';

const Duration _pingInterval = Duration(seconds: 15);
const Duration _staleLimit = Duration(seconds: 35);
const Duration _watchdogInterval = Duration(seconds: 5);
const Duration _resyncDebounce = Duration(seconds: 2);
const Duration _minBackoff = Duration(milliseconds: 500);
const Duration _maxBackoff = Duration(seconds: 15);

/// Sent on the upgrade so the relay can recognise a native client. The relay
/// treats the Origin check as browser CSRF hygiene, not authentication, and the
/// operator opts in by listing this value in ALLOWED_ORIGINS.
const String appOrigin = 'app://herebee';

class NetHandlers {
  const NetHandlers({
    required this.onPeer,
    required this.onLeft,
    required this.onRequest,
    required this.onStatus,
    required this.onPresence,
    required this.onFatal,
  });

  final void Function(String id, PeerUpdate update) onPeer;
  final void Function(String id) onLeft;

  /// A peer joined; re-broadcast our latest state.
  final void Function() onRequest;
  final void Function(bool connected) onStatus;

  /// Room occupancy (count only, no identity).
  final void Function(int n) onPresence;
  final void Function(String reason) onFatal;
}

class NetClient {
  NetClient({
    required this.endpoint,
    required this.keys,
    required this.cid,
    required this.clientLabel,
    required this.handlers,
  });

  /// e.g. wss://herebee.app/ws
  final String endpoint;
  final RoomKeys keys;
  final String cid;

  /// `<platform>/<version>`, sent as X-HereBee-Client.
  final String clientLabel;
  final NetHandlers handlers;

  WebSocket? _ws;
  StreamSubscription<dynamic>? _sub;
  Duration _backoff = _minBackoff;
  bool _closed = false;
  DateTime _lastResync = DateTime.fromMillisecondsSinceEpoch(0);

  /// The last "loc" frame, replayed on reconnect so peers see us again.
  String? _lastSent;
  DateTime _lastActivity = DateTime.fromMillisecondsSinceEpoch(0);
  Timer? _pingTimer;
  Timer? _watchdog;
  Timer? _reconnectTimer;

  Future<void> connect() async {
    if (_closed) return;
    // Always start from a clean slate: a previous socket may be dead but still
    // open (standby zombie), so a reconnect must never stack a second socket on
    // a live one.
    await _teardown();

    final WebSocket ws;
    try {
      ws = await WebSocket.connect(
        endpoint,
        headers: {'Origin': appOrigin, 'X-HereBee-Client': clientLabel},
      );
    } catch (_) {
      handlers.onStatus(false);
      _scheduleReconnect();
      return;
    }
    if (_closed) {
      await ws.close();
      return;
    }

    _ws = ws;
    _backoff = _minBackoff;
    _lastActivity = DateTime.now();
    _send(joinFrame(keys.roomId, cid));
    handlers.onStatus(true);
    if (_lastSent != null) ws.add(_lastSent!); // resume visibility after reconnect
    _startHeartbeat();

    _sub = ws.listen(
      _onFrame,
      onDone: _onDone,
      onError: (Object _) => _onDone(),
      cancelOnError: false,
    );
  }

  Future<void> _onFrame(dynamic raw) async {
    _lastActivity = DateTime.now(); // any inbound frame proves the link is alive
    if (raw is! String) return;
    Object? decoded;
    try {
      decoded = jsonDecode(raw);
    } catch (_) {
      return;
    }
    final msg = ServerMessage.parse(decoded);
    switch (msg) {
      case null:
      case HelloMessage():
      case PongMessage():
        return; // liveness / ack only — activity already recorded above
      case PeerMessage(:final id, :final data):
        final update = validPeerUpdate(await decryptJson(keys.key, data));
        if (update != null) handlers.onPeer(id, update);
      case LeftMessage(:final id):
        handlers.onLeft(id);
      case RequestMessage():
        handlers.onRequest();
      case PresenceMessage(:final n):
        handlers.onPresence(n);
      case ErrorMessage(:final reason):
        _closed = true;
        handlers.onFatal(reason);
    }
  }

  void _onDone() {
    _stopHeartbeat();
    handlers.onStatus(false);
    if (!_closed) _scheduleReconnect();
  }

  void _send(Map<String, dynamic> frame) => _ws?.add(jsonEncode(frame));

  Future<void> _teardown() async {
    _stopHeartbeat();
    final sub = _sub;
    final ws = _ws;
    _sub = null;
    _ws = null;
    await sub?.cancel();
    try {
      await ws?.close();
    } catch (_) {
      /* already gone */
    }
  }

  /// Application-level liveness. A dropped network can leave the socket wedged
  /// open, so `onDone` may never fire. Ping periodically and, if nothing arrives
  /// within the stale limit, rebuild the connection.
  void _startHeartbeat() {
    _stopHeartbeat();
    _lastActivity = DateTime.now();
    _pingTimer = Timer.periodic(_pingInterval, (_) => _send(pingFrame));
    _watchdog = Timer.periodic(_watchdogInterval, (_) {
      if (_ws == null) return;
      if (DateTime.now().difference(_lastActivity) <= _staleLimit) return;
      handlers.onStatus(false);
      unawaited(connect()); // tears down the zombie and reconnects
    });
  }

  void _stopHeartbeat() {
    _pingTimer?.cancel();
    _pingTimer = null;
    _watchdog?.cancel();
    _watchdog = null;
  }

  void _scheduleReconnect() {
    _reconnectTimer?.cancel();
    final delay = _backoff;
    _backoff = Duration(
      milliseconds: (_backoff.inMilliseconds * 2).clamp(0, _maxBackoff.inMilliseconds),
    );
    _reconnectTimer = Timer(delay, () {
      if (!_closed) unawaited(connect());
    });
  }

  /// Force a fresh connection. Call this when the app returns to the foreground:
  /// standby often leaves a socket that still reports open but carries nothing.
  void resync() {
    if (_closed) return;
    final now = DateTime.now();
    if (now.difference(_lastResync) < _resyncDebounce) return;
    _lastResync = now;
    _backoff = _minBackoff;
    unawaited(connect());
  }

  Future<void> broadcast(PeerUpdate update) async {
    final data = await encryptJson(keys.key, update.toJson());
    final frame = jsonEncode(relayFrame(data));
    // Replayed on every reconnect. A "stop" must clear it, or a later reconnect
    // would resurrect our old position and make us reappear after we stopped.
    if (update is LocUpdate) {
      _lastSent = frame;
    } else {
      _lastSent = null;
    }
    _ws?.add(frame);
  }

  Future<void> close() async {
    _closed = true;
    _reconnectTimer?.cancel();
    await _teardown();
  }
}
